overlord/devicestate/devicestate.go

```diff
@@ -709,6 +709,11 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 	// go through all the model snaps, see if there are new required snaps
 	// or a track for existing ones needs to be updated
 	for _, modelSnap := range new.SnapsWithoutEssential() {
+		// if the snap isn't required, then we don't do anything right now
+		if modelSnap.Presence != "required" {
```

Comment:
If the presence wasn't required, then `needsInstall` was not set to true. This resulted in the `*state.TaskSet (ts)` that is used in this for-loop being `nil`. On line 770, we use `ts`, causing a panic.

I think this for-loop could use some refactoring to prevent bugs like this. I have a change ready that I think will make a mistake like this less likely. However, I wanted to make the actual bug easier to review.